tests/unit/nupic/support/decorators_test.py

```diff
@@ -22,18 +22,251 @@
+
+    _retry = decorators.retry(
+      timeoutSec=0, initialRetryDelaySec=0.2,
+      maxRetryDelaySec=10)
+
+    testFunction = Mock(side_effect=TestParentException("Test exception"),
+                        __name__="testFunction", autospec=True)
+
+    with self.assertRaises(TestParentException):
+      _retry(testFunction)()
+
+    self.assertFalse(mockSleep.called)
+    testFunction.assert_called_once_with()
+
+
+  @patch("time.sleep", autospec=True)
+  @patch("time.time", autospec=True)
+  def testRetryWaitsInitialRetryDelaySec(self, mockTime, mockSleep):
+    """Test that delay times are correct."""
+
+    self.mockSleepTime(mockTime, mockSleep)
+
+    _retry = decorators.retry(
+      timeoutSec=30, initialRetryDelaySec=2,
+      maxRetryDelaySec=10)
+
+    testFunction = Mock(side_effect=TestParentException("Test exception"),
+                        __name__="testFunction", autospec=True)
+
+    with self.assertRaises(TestParentException):
+      _retry(testFunction)()
+
+    self.assertEqual(mockSleep.mock_calls, [call(2), call(4), call(8),
+                                            call(10), call(10)])
+
+    self.assertEqual(testFunction.call_count, 6)
+
+
+  @patch("time.sleep", autospec=True)
+  @patch("time.time", autospec=True)
+  def testRetryRetryExceptionIncluded(self, mockTime, mockSleep):
+    """Test that retry is triggered if raised exception is in
+    retryExceptions."""
+
+    self.mockSleepTime(mockTime, mockSleep)
+
+    _retry = decorators.retry(
+      timeoutSec=1, initialRetryDelaySec=1,
+      maxRetryDelaySec=10, retryExceptions=(TestParentException,))
+
+    @_retry
+    def testFunction():
+      raise TestChildException("Test exception")
+
+    with self.assertRaises(TestChildException):
+      testFunction()
+
+    self.assertEqual(mockSleep.call_count, 1)
+
+
+  @patch("time.sleep", autospec=True)
+  @patch("time.time", autospec=True)
+  def testRetryRetryExceptionExcluded(self, mockTime, mockSleep):
+    """ Test that retry is not triggered if raised exeception is not in
+    retryExceptions """
+
+    self.mockSleepTime(mockTime, mockSleep)
+
+    class TestExceptionA(Exception):
+      pass
+
+    class TestExceptionB(Exception):
+      pass
+
+    _retry = decorators.retry(
+      timeoutSec=1, initialRetryDelaySec=1,
+      maxRetryDelaySec=10, retryExceptions=(TestExceptionA,))
+
+    @_retry
+    def testFunction():
+      raise TestExceptionB("Test exception")
+
+    with self.assertRaises(TestExceptionB):
+      testFunction()
+
+    self.assertEqual(mockSleep.call_count, 0)
+
+
+  @patch("time.sleep", autospec=True)
+  @patch("time.time", autospec=True)
+  def testRetryRetryFilter(self, mockTime, mockSleep):
+    """Test that if retryFilter is specified and exception is in
+    retryExceptions, retries iff retryFilter returns true."""
+
+    self.mockSleepTime(mockTime, mockSleep)
+
+    # Test with retryFilter returning True
+
+    _retryTrueFilter = decorators.retry(
```

Comment:
Shouldn't be using the underscore in the name `_retryTrueFilter`. I see more of this in other local variables. Please fix them all.